tests/model_specific/llama_cpp/test_chat_templates.py

```diff
@@ -0,0 +1,46 @@
+import jinja2
+import pytest
+
+import guidance
+
+from guidance.chat import CHAT_TEMPLATE_CACHE
+
+
+def test_chat_format_smoke(llamacpp_model: guidance.models.LlamaCpp):
+    # Retrieve the template string
+    if (
+        hasattr(llamacpp_model.engine.model_obj, "metadata")
+        and "tokenizer.chat_template" in llamacpp_model.engine.model_obj.metadata
+    ):
+        model_chat_template = llamacpp_model.engine.model_obj.metadata["tokenizer.chat_template"]
+    else:
+        pytest.skip("Chat template not available from LlamaCpp object")
+
+    lm = guidance.models.Mock("")
+    lm.chat_template = CHAT_TEMPLATE_CACHE[model_chat_template]()
+
+    messages = [
+        {"role": "user", "content": "Good_day_to_you!"},
+        {"role": "assistant", "content": "Hello!"},
+    ]
+
+    # Note that llama-cpp-python does provide a llama_chat_apply_template function
+    # but details about its use are thin on the ground and according to
+    # https://github.com/ggerganov/llama.cpp/wiki/Templates-supported-by-llama_chat_apply_template
+    # it does its own thing internally
+    jinja2_template = jinja2.Environment(loader=jinja2.BaseLoader()).from_string(
+        model_chat_template
+    )
+    jinja2_render = jinja2_template.render(
+        messages=messages,
+        bos_token=llamacpp_model.engine.tokenizer.bos_token.decode(),
+        eos_token=llamacpp_model.engine.tokenizer.eos_token.decode(),
+    )
+
+    with guidance.user():
+        lm += "Good_day_to_you!"
+    with guidance.assistant():
+        lm += "Hello!"
+    # Only check substring due to BOS/EOS tokens
+    # Remove spaces because those are being troublesome as well
+    assert str(lm).replace(" ", "") in jinja2_render.replace(" ", "")
```

Comment:
It was in Llama2, I was trying to use the same template class as for Transformers. The spacing around the 'assistant' tags was different.

Let me check a couple more things, but the Jinja2 strings have quite a few differences between the Transformers and LlamaCpp versions of Llama2

Reply:
So, to be precise, for `hfllama_mistral_7b`, we have:
```
str(lm)       = '<s>[INST] Good_day_to_you! [/INST] Hello! </s>'
jinja2_render = '<s>[INST] Good_day_to_you! [/INST]Hello!</s>'
```
So there are extra spaces inserted by our template, rather than in the Jinja2 template being extracted from the loaded GGUF. I would expect these two be benign, but I obviously can't prove that, and it doesn't help the test pass.

We could drop the BOS/EOS tokens from the match, but that would still leave the extra space after `[/INST]` which our `ChatTemplate` is inserting (and matches Mistral when loaded from Transformers).